File: frontend/src/utils/featureFlags.ts
/**
 * Feature flags for managing 50+ features.
 *
 * Usage:
 *   import { features } from "@/utils/featureFlags";
 *   if (features.dsaVisualizer) { ... }
 */

export type FeatureFlag = {
  key: string;
  enabled: boolean;
  description: string;
  rolloutPercent?: number;
  allowedPlans?: ("free" | "pro" | "lifetime")[];
};

export const features: Record<string, FeatureFlag> = {
  // Core features (always on)
  interviews: { key: "interviews", enabled: true, description: "AI mock interviews" },
  resume: { key: "resume", enabled: true, description: "Resume builder and analyzer" },
  aptitude: { key: "aptitude", enabled: true, description: "Aptitude test practice" },
  coding: { key: "coding", enabled: true, description: "Coding challenges" },
  compiler: { key: "compiler", enabled: true, description: "LeetCode-style compiler" },
  questions: { key: "questions", enabled: true, description: "Question bank" },

  // AI features
  aiMentor: { key: "aiMentor", enabled: true, description: "AI mentor chat" },
  projectGenerator: { key: "projectGenerator", enabled: true, description: "AI project generator" },
  coverLetter: { key: "coverLetter", enabled: true, description: "Cover letter generator" },
  salaryNegotiation: { key: "salaryNegotiation", enabled: true, description: "Salary negotiation coach" },
  systemDesign: { key: "systemDesign", enabled: true, description: "System design practice" },
  companyPrep: { key: "companyPrep", enabled: true, description: "Company-specific prep" },
  aiDebugger: { key: "aiDebugger", enabled: true, description: "AI-powered debugger" },
  aiFeedback: { key: "aiFeedback", enabled: true, description: "Real-time AI feedback" },

  // Gamification
  gamification: { key: "gamification", enabled: true, description: "XP, levels, streaks, badges" },
  tower: { key: "tower", enabled: true, description: "Placement tower progression" },
  battlePass: { key: "battlePass", enabled: true, description: "Battle pass system" },
  dailyChallenges: { key: "dailyChallenges", enabled: true, description: "Daily adaptive challenges" },
  mysteryBoxes: { key: "mysteryBoxes", enabled: true, description: "Mystery box rewards" },
  powerUps: { key: "powerUps", enabled: true, description: "Power-up shop" },
  bossBattles: { key: "bossBattles", enabled: true, description: "Boss battles at levels 10,20..." },
  leaderboard: { key: "leaderboard", enabled: true, description: "Global leaderboard" },
  rank: { key: "rank", enabled: true, description: "Honor/Kyu-Dan rank system" },

  // Social/Community
  community: { key: "community", enabled: true, description: "Community discussions" },
  studyGroups: { key: "studyGroups", enabled: true, description: "Study groups" },
  battles: { key: "battles", enabled: true, description: "1v1 coding battles" },
  tournaments: { key: "tournaments", enabled: true, description: "Monthly contests" },
  scrims: { key: "scrims", enabled: true, description: "Scrimba-style screencasts" },
  chat: { key: "chat", enabled: true, description: "Real-time chat" },
  referrals: { key: "referrals", enabled: true, description: "Referral program" },

  // Learning
  learningHub: { key: "learningHub", enabled: true, description: "Learning hub" },
  languagePaths: { key: "languagePaths", enabled: true, description: "7 language learning paths" },
  learningModules: { key: "learningModules", enabled: true, description: "Duolingo-style lessons" },
  adaptivePath: { key: "adaptivePath", enabled: true, description: "AI-driven adaptive learning" },
  dsaFingerprint: { key: "dsaFingerprint", enabled: true, description: "DSA skill assessment" },
  dsaVisualizer: { key: "dsaVisualizer", enabled: true, description: "Algorithm visualizations" },
  concepts: { key: "concepts", enabled: true, description: "Concept explanations" },
  spacedRepetition: { key: "spacedRepetition", enabled: true, description: "SRS mastery system" },

  // Career tools
  atsOptimizer: { key: "atsOptimizer", enabled: true, description: "ATS score optimizer" },
  salaryBenchmark: { key: "salaryBenchmark", enabled: true, description: "Salary benchmark data" },
  applicationTracker: { key: "applicationTracker", enabled: true, description: "Job application tracker" },
  careerProfile: { key: "careerProfile", enabled: true, description: "Career profile" },
  interviewBooking: { key: "interviewBooking", enabled: true, description: "Interview booking system" },
  mockOA: { key: "mockOA", enabled: true, description: "Mock online assessments" },
  predictor: { key: "predictor", enabled: true, description: "Placement prediction engine" },
  readiness: { key: "readiness", enabled: true, description: "Interview readiness scoring" },

  // Indian market
  indianPlacement: { key: "indianPlacement", enabled: true, description: "Indian placement prep" },
  campusConnect: { key: "campusConnect", enabled: true, description: "Campus connect" },
  campusWars: { key: "campusWars", enabled: true, description: "Campus wars leaderboard" },
  placementDrives: { key: "placementDrives", enabled: true, description: "Placement drives" },
  alumniExperiences: { key: "alumniExperiences", enabled: true, description: "Alumni experiences" },

  // Advanced features
  worldMap: { key: "worldMap", enabled: true, description: "World map exploration" },
  skillTrees: { key: "skillTrees", enabled: true, description: "Visual skill trees" },
  dungeons: { key: "dungeons", enabled: true, description: "Dungeon challenges" },
  guilds: { key: "guilds", enabled: true, description: "Guild system" },
  economy: { key: "economy", enabled: true, description: "Player economy" },
  seasons: { key: "seasons", enabled: true, description: "Seasonal events" },
  gameEvents: { key: "gameEvents", enabled: true, description: "Game events" },
  cardCollection: { key: "cardCollection", enabled: true, description: "Card collection" },
  showcase: { key: "showcase", enabled: true, description: "Project showcase gallery" },
  timelines: { key: "timelines", enabled: true, description: "Placement timeline" },

  // Admin/Enterprise
  adminDashboard: { key: "adminDashboard", enabled: true, description: "Admin analytics dashboard", allowedPlans: ["pro", "lifetime"] },
  adminContent: { key: "adminContent", enabled: true, description: "Admin content management", allowedPlans: ["pro", "lifetime"] },
  enterprise: { key: "enterprise", enabled: true, description: "Enterprise plan features", allowedPlans: ["pro", "lifetime"] },
  billing: { key: "billing", enabled: true, description: "PayPal/Stripe billing" },

  // Experimental
  pwa: { key: "pwa", enabled: true, description: "Progressive Web App" },
  pushNotifications: { key: "pushNotifications", enabled: true, description: "Push notifications" },
  offlineMode: { key: "offlineMode", enabled: true, description: "Offline mode" },
  steamProfile: { key: "steamProfile", enabled: true, description: "Steam-style profile" },
  newspaper: { key: "newspaper", enabled: true, description: "Placement Times newspaper" },
  luckyWheel: { key: "luckyWheel", enabled: true, description: "Daily lucky wheel" },
  retentionAdmin: { key: "retentionAdmin", enabled: true, description: "Retention analytics admin", allowedPlans: ["pro", "lifetime"] },
};

export function isFeatureEnabled(flagKey: string, userPlan: string = "free"): boolean {
  const flag = features[flagKey];
  if (!flag) return false;
  if (!flag.enabled) return false;

  if (flag.allowedPlans && !flag.allowedPlans.includes(userPlan as "free" | "pro" | "lifetime")) {
    return false;
  }

  if (flag.rolloutPercent !== undefined) {
    return Math.random() * 100 < flag.rolloutPercent;
  }

  return true;
}

export function getEnabledFeatures(userPlan: string = "free"): string[] {
  return Object.entries(features)
    .filter(([key, _]) => isFeatureEnabled(key, userPlan))
    .map(([key]) => key);
}

export function setFeatureEnabled(key: string, enabled: boolean): void {
  if (features[key]) {
    features[key].enabled = enabled;
  }
}

File: backend/app/services/feature_flags.py
"""Feature flags for managing 50+ features with gradual rollout support."""

from __future__ import annotations

from typing import Literal

Plan = Literal["free", "pro", "lifetime"]

from enum import Enum


class FeatureStatus(str, Enum):
    """Status of a feature flag."""
    DISABLED = "disabled"
    INTERNAL = "internal"
    BETA = "beta"
    GRADUAL = "gradual"
    ENABLED = "enabled"


class FeatureManager:
    """Simple feature flag manager for admin operations."""

    def __init__(self, flags: dict | None = None) -> None:
        self.flags = flags or FEATURE_FLAGS

    def get_all_flags(self) -> dict:
        """Return all feature flags."""
        return self.flags

    def update_flag(
        self,
        key: str,
        enabled: bool | None = None,
        status: FeatureStatus | None = None,
        rollout_percentage: int | None = None,
    ) -> None:
        """Update a feature flag."""
        if key not in self.flags:
            return
        flag = self.flags[key]
        if enabled is not None:
            flag["enabled"] = enabled
        if status is not None:
            flag["status"] = status.value
        if rollout_percentage is not None:
            flag["rollout_percentage"] = rollout_percentage

FEATURE_FLAGS: dict[str, dict] = {
    "interviews": {"enabled": True, "description": "AI mock interviews"},
    "resume": {"enabled": True, "description": "Resume builder and analyzer"},
    "aptitude": {"enabled": True, "description": "Aptitude test practice"},
    "coding": {"enabled": True, "description": "Coding challenges"},
    "compiler": {"enabled": True, "description": "LeetCode-style compiler"},
    "questions": {"enabled": True, "description": "Question bank"},
    "ai_mentor": {"enabled": True, "description": "AI mentor chat"},
    "project_generator": {"enabled": True, "description": "AI project generator"},
    "cover_letter": {"enabled": True, "description": "Cover letter generator"},
    "salary_negotiation": {"enabled": True, "description": "Salary negotiation coach"},
    "system_design": {"enabled": True, "description": "System design practice"},
    "company_prep": {"enabled": True, "description": "Company-specific prep"},
    "ai_debugger": {"enabled": True, "description": "AI-powered debugger"},
    "ai_feedback": {"enabled": True, "description": "Real-time AI feedback"},
    "gamification": {"enabled": True, "description": "XP, levels, streaks, badges"},
    "tower": {"enabled": True, "description": "Placement tower progression"},
    "daily_challenges": {"enabled": True, "description": "Daily adaptive challenges"},
    "leaderboard": {"enabled": True, "description": "Global leaderboard"},
    "rank": {"enabled": True, "description": "Honor/Kyu-Dan rank system"},
    "community": {"enabled": True, "description": "Community discussions"},
    "learning_hub": {"enabled": True, "description": "Learning hub"},
    "language_paths": {"enabled": True, "description": "7 language learning paths"},
    "learning_modules": {"enabled": True, "description": "Duolingo-style lessons"},
    "adaptive_path": {"enabled": True, "description": "AI-driven adaptive learning"},
    "dsa_fingerprint": {"enabled": True, "description": "DSA skill assessment"},
    "dsa_visualizer": {"enabled": True, "description": "Algorithm visualizations"},
    "concepts": {"enabled": True, "description": "Concept explanations"},
    "spaced_repetition": {"enabled": True, "description": "SRS mastery system"},
    "ats_optimizer": {"enabled": True, "description": "ATS score optimizer"},
    "salary_benchmark": {"enabled": True, "description": "Salary benchmark data"},
    "application_tracker": {"enabled": True, "description": "Job application tracker"},
    "career_profile": {"enabled": True, "description": "Career profile"},
    "interview_booking": {"enabled": True, "description": "Interview booking system"},
    "mock_oa": {"enabled": True, "description": "Mock online assessments"},
    "predictor": {"enabled": True, "description": "Placement prediction engine"},
    "readiness": {"enabled": True, "description": "Interview readiness scoring"},
    "indian_placement": {"enabled": True, "description": "Indian placement prep"},
    "campus_connect": {"enabled": True, "description": "Campus connect"},
    "campus_wars": {"enabled": True, "description": "Campus wars leaderboard"},
    "placement_drives": {"enabled": True, "description": "Placement drives"},
    "alumni_experiences": {"enabled": True, "description": "Alumni experiences"},

    "admin_dashboard": {"enabled": True, "description": "Admin analytics dashboard", "allowed_plans": ["pro", "lifetime"]},
    "admin_content": {"enabled": True, "description": "Admin content management", "allowed_plans": ["pro", "lifetime"]},
    "enterprise": {"enabled": True, "description": "Enterprise plan features", "allowed_plans": ["pro", "lifetime"]},
    "billing": {"enabled": True, "description": "PayPal/Stripe billing"},
    "pwa": {"enabled": True, "description": "Progressive Web App"},
    "push_notifications": {"enabled": True, "description": "Push notifications"},
    "offline_mode": {"enabled": True, "description": "Offline mode"},

    "retention_admin": {"enabled": True, "description": "Retention analytics admin", "allowed_plans": ["pro", "lifetime"]},
}


def is_feature_enabled(flag_key: str, plan: Plan = "free") -> bool:
    """Check if a feature is enabled for the given plan."""
    flag = FEATURE_FLAGS.get(flag_key)
    if not flag or not flag.get("enabled", False):
        return False

    allowed_plans = flag.get("allowed_plans")
    if allowed_plans and plan not in allowed_plans:
        return False

    return True


def get_enabled_features(plan: Plan = "free") -> list[str]:
    """Get all feature keys enabled for the given plan."""
    return [key for key, flag in FEATURE_FLAGS.items() if is_feature_enabled(key, plan)]


def set_feature_enabled(key: str, enabled: bool) -> None:
    """Enable or disable a feature flag."""
    if key in FEATURE_FLAGS:
        FEATURE_FLAGS[key]["enabled"] = enabled


async def init_feature_flags(db) -> None:
    """Initialize the feature-flag system.

    Flags are stored in-memory in ``FEATURE_FLAGS`` (see module docstring).
    This initializer is intentionally a no-op so the service can be dropped in
    as a startup hook without a DB dependency; future DB-backed rollout will go here.
    """
    return None


def get_feature_manager() -> FeatureManager:
    """Get a feature flag manager instance."""
    return FeatureManager()
